### PR TITLE
Distribution analysis

### PR DESCRIPTION
**Pull request by konstantgr**
Added final version of clone analysis

### CODE REVIEW

**Comment by TitovS** [comment]
General comments:
- test poetry one more time -- I had some problems, but poetry update helped. 
- please, hash scripts names
- typing is important, please add it
- Let's not keep in repo archives or add functionalty to unpack it.

File: experiment.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        self.files = self._get_files()
+        self.in_path = in_path
+        self.source_files = self._get_source_files() if in_path else {'notebooks': [], 'scripts': []}
+
+        self.length_range = None
+        self.aggregated_stats = {'notebooks': None, 'scripts': None}
+
+    @staticmethod
+    def _get_json_files(folder, max_num):
```
Comment (TitovS) [inline experiment.py]: is it same max_num as in class? Why is it static? Same for folder

File: experiment.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        return [folder / Path(pos_json) for pos_json in os.listdir(folder) if
```
Comment (TitovS) [inline experiment.py]: We are switching to pathlib because of its superiority. Please use Path.iterdir()

File: experiment.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                pos_json.endswith('.json')][:max_num]
+
+    def _get_source_files(self):
+        return {k: [get_source_path(file, in_folder=folder) for file in self.files[k]]
+                for k, folder in self.in_path.items()}
```
Comment (TitovS) [inline experiment.py]: I didn't get the idea -- what the functionality of this piece?

File: experiment.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _get_files(self):
+        notebook_files = self._get_json_files(self.notebooks_folder, self.max_num)
+        scripts_files = None if not self.scripts_folder else self._get_json_files(self.scripts_folder, self.max_num)
```
Comment (TitovS) [inline experiment.py]: May be we can make both variable None if there is no folder + check with error if both are none?

File: utils/processing_utils.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        for i, d_i in d.items():
+            try:
+                d[i] = json.loads(d_i)
+            except (json.decoder.JSONDecodeError, TypeError):
+                continue
+
+        return d
+
+
+def filter_clones(
```
Comment (TitovS) [inline utils/processing_utils.py]: Why this method is in utils. It is definitely part of experiment.

File: utils/processing_utils.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import json
+from typing import List, Dict
+from pathlib import Path
+
+
+def read_clones_data(filepath: Path) -> Dict:
+    with open(filepath, 'r') as f:
```
Comment (TitovS) [inline utils/processing_utils.py]: Please use Pathlib functionality. See path.open for details.

File: utils/processing_utils.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        d = json.load(f)
+
+        for i, d_i in d.items():
+            try:
+                d[i] = json.loads(d_i)
+            except (json.decoder.JSONDecodeError, TypeError):
```
Comment (TitovS) [inline utils/processing_utils.py]: Please better naming

File: utils/processing_utils.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                continue
```
Comment (TitovS) [inline utils/processing_utils.py]: Never use try-except construction without any type of logging. Add print() at least.

File: utils/processing_utils.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def get_stats(lst: List, norm: int = 1) -> Dict:
+    return {
+        'groups_cnt': len(lst) / norm,
+        'clones_cnt': sum([len(g["clones"]) for g in lst]) / norm
+    }
+
+
+def is_break(start: int, finish: int, source: str) -> bool:
```
Comment (TitovS) [inline utils/processing_utils.py]: it is also part of experiment. For example, sep should be class constant.

File: utils/processing_utils.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        data["3"]["groups"]
+    ))
+    if breaks:
+        for i, g in enumerate(lst):
+            lst[i]['clones'] = filter_breaks(g['clones'], source_path)
+
+    return lst
+
+
+def get_stats(lst: List, norm: int = 1) -> Dict:
```
Comment (TitovS) [inline utils/processing_utils.py]: Why not stat util?

File: experiment.py (hunk 3 above)
Comment (TitovS) [reply experiment.py]: Finally, got it. Let's discuss it on call -- we should redo this.

File: experiment.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+                path = source_paths[i] if source_paths else None
+                drop_breaks = drop_breaks if path else None
+
+                stats_tmp = [
+                    get_stats(
+                        filter_clones(data=data_tmp, min_length=min_l, breaks=drop_breaks, source_path=path),
+                        norm=norm
+                    )
+                    for min_l in length_range
+                ]
```
Comment (TitovS) [inline experiment.py]: While i really appreciate list comprehensions, this is too complicated -- lest refactor this to the start loop

File: experiment.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        files = {'notebooks': notebook_files, 'scripts': scripts_files}
+        return files
+
+    @staticmethod
+    def aggregate(files, length_range, source_paths, normalize=False, drop_breaks=False):
+        stats = []
+
+        for i, file in tqdm(enumerate(files)):
+            data_tmp = read_clones_data(file)
+            try:
```
Comment (TitovS) [inline experiment.py]: Please, decrease size of this try construction. There are only a few lines where KeyError is possible.

File: experiment.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+                        norm=norm
+                    )
+                    for min_l in length_range
+                ]
+                stats_tmp = pd.DataFrame(stats_tmp)
+                stats_tmp['min_length'] = list(length_range)
+
+                stats.append(pd.DataFrame(stats_tmp))
+            except KeyError:
+                continue
```
Comment (TitovS) [inline experiment.py:89]: At least add print()

File: experiment.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                    get_stats(
+                        filter_clones(data=data_tmp, min_length=min_l, breaks=drop_breaks, source_path=path),
+                        norm=norm
+                    )
+                    for min_l in length_range
+                ]
+                stats_tmp = pd.DataFrame(stats_tmp)
+                stats_tmp['min_length'] = list(length_range)
+
+                stats.append(pd.DataFrame(stats_tmp))
```
Comment (TitovS) [inline experiment.py]: Redundant constructor?

File: experiment.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+        ax.set_title(f"Clones count {stats_type}")
+        ax.set_xlabel("Min clone length")
+        ax.set_ylabel("Clones count")
+
+        if params and params.get('log'):
+            ax.set_yscale('log')
+            ax.set_ylabel("Clones count (log)")
+
+        return ax
```
Comment (TitovS) [inline experiment.py]: and this could be extracted to the utils.

File: utils/data_utils.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    data = []
+    for path in base_path.rglob('*.json'):
+        try:
+            with open(path, 'r') as f:
+                d = json.loads(json.load(f).get('3'))
+                d['name'] = path.name
+                data.append(d)
+
+        except TypeError:
+            continue
```
Comment (TitovS) [inline utils/data_utils.py:36]: log errors!

**Review by TitovS — CHANGES_REQUESTED** [review]

File: experiment.py (hunk 2 above)
Comment (konstantgr) [reply experiment.py]: Ok. Will use rglob

File: experiment.py (hunk 1 above)
Comment (konstantgr) [reply experiment.py]: done

File: experiment.py (hunk 2 above)
Comment (konstantgr) [reply experiment.py]: done

File: utils/processing_utils.py (hunk 5 above)
Comment (konstantgr) [reply utils/processing_utils.py]: done

File: experiment.py (hunk 12 above)
Comment (konstantgr) [reply experiment.py]: done

File: experiment.py (hunk 13 above)
Comment (konstantgr) [reply experiment.py:89]: done

File: experiment.py (hunk 15 above)
Comment (konstantgr) [reply experiment.py]: deleted

File: utils/data_utils.py (hunk 16 above)
Comment (konstantgr) [reply utils/data_utils.py:36]: done

**Pull request merged**